zokrates_stdlib/stdlib/signatures/verifyEddsa.code

```diff
@@ -0,0 +1,47 @@
+import "hashes/sha256/1024bitPadded.code" as sha256
+import "ecc/edwardsScalarMult.code" as scalarMult 
+import "ecc/edwardsAdd.code" as add
+import "utils/pack/unpack256.code" as unpack256
+import "ecc/edwardsOnCurve.code" as onCurve
+import "ecc/edwardsOrderCheck.code" as orderCheck
+
+/// Verifies an EdDSA Signature.
+///
+/// Checks the correctness of a given EdDSA Signature (R,S) for the provided
+/// public key A and message (M0, M1).
+/// For more information see:
+/// https://en.wikipedia.org/wiki/EdDSA
+/// https://eprint.iacr.org/2015/677.pdf
+///
+/// Arguments:
+///    R: Curve point. Hidden version of the per-message nonce.
+///    S: Field element. Signature to be verified.
+///    A: Curve point. Public part of the key used to create S.
+///    M0: 256bit array. First 256bits of the message used to create S  .
+///    M1: 256bit array. Trailing 256bits of the message used to create S  .
+///    context: Curve parameters used to create S.
+///
+/// Returns:
+///     Return 1 for S being a valid EdDSA Signature, 0 otherwise.
+def main(private field[2] R, private field S, field[2] A, field[256] M0, field[256] M1, field[10] context) -> (field):
+
+    field[2] G = [context[4], context[5]]
+
+    // Check if R is on curve and in prime-order sub-group. A is public input and can be checked offline 
+    field isOnCurve = onCurve(R, context) // throws if R is not on curve
+    field isPrimeOrder = orderCheck(R, context)
```

Comment:
This doesn't check it's in the prime order subgroup, it checks that it's not in the small subgroup. It could be in a coset, i.e., P = a*G + b*H where G is the generator of the prime subgroup and H is the generator of the small subgroup, and the check would still pass.

The check itself is fine, just not the comment :)

Additionally, verification in edDSA multiplies by the cofactor in the verification equation (why I think you tried to avoid by checking you're in the prime subgroup).

It might not be completely necessary, but it deviates from https://tools.ietf.org/html/rfc8032#page-43 and librustzcash, for example.

Reply:
Thanks a lot for pointing that out. I will update the comment accordingly.

Regarding the cofactorless verification this implementation follows:
https://eprint.iacr.org/2015/677.pdf  (see section "Cofactorless  verification")
We've decided to go that way to be compatible with the work in
- ethsnarks: https://github.com/HarryR/ethsnarks/blob/master/ethsnarks/eddsa.py#L193
- thematter: https://github.com/matter-labs/sapling-crypto/blob/master/src/circuit/baby_eddsa.rs#L121